### PR TITLE
Task #286 - Chapter for starting a new project

### PR DESCRIPTION
**Pull request by nikajukic**
Task: [#286](https://app.productive.io/1-infinum/task/1043730)

#### Aim
A chapter for things you need to do when you're starting a new project.

### CODE REVIEW

File: Starting a new project.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+If you need any other services that aren't mentioned, **provide as many information as possible**
+
+**Project setup sheet needs to be up to date** with new changes to project setup and services used. If you're adding some new dependency or service or changing something, please update it in project setup sheet and notify DevOps to make these changes.
+
+
+#### 3. Generate new application
+
+Use [**default rails template**](https://github.com/infinum/default_rails_template/) to start your new application.
+
+Most of our work is based on the Rails framework. We have some standard configurations and gems that are basically in every Rails application. To make our lives easier, we've made a default template to generate new Rails applications that meet those standards.
```
Comment (melcha) [inline Starting a new project.md:28]: I'd add here: 

Please always update the default rails template if something gets broken or out of date when you're starting a new project.

File: Starting a new project.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+#### 5. Fill out readme
+
+After setting up a project, fill out a readme file with all the necessary details about the application:
+* dependencies & prerequisites
+* setup instructions
+* development instructions
+* deployment instructions
+* test instructions
+* any other project specific information that someone who takes over a project needs to know
```
Comment (melcha) [inline Starting a new project.md:51]: The ER diagram should also be saved somewhere in the repository, maybe a separate file, and it should be linked from the readme.

**Comment by melcha** [comment]
@infinum/backend-team please read this and comment if something else should be added to this checklist.

File: Starting a new project.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+#### 2. Fill out Project setup sheet
+
+We've come up with a project setup template in cooperation with DevOps team. It's purpose is to provide the DevOps with all the details needed for deploying your application to staging and production servers.
+
+It contains general info about your project that your project manager can fill out as well as more technical information like frameworks and technologies used, build and deploy setup, third party services that need to be set up...
+
+Ask your PM for a link to the sheet template.
+
+**Fill in the blanks on Backend related stuff.**
```
Comment (nikajukic) [inline Starting a new project.md:17]: Just one comment here: We're creating a new project setup template and it's going to have a separate sheet for backend related stuff.

File: Starting a new project.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
+## Starting a new project
+
+When you're starting a new project, there are a few things you need to do:
+
+#### 1. Open a repository
+
+Ask your team lead or DevOps to open a repository.
+
+#### 2. Fill out Project setup sheet
```
[the remark this answers is not in the document]
Comment (cukoreva) [reply Starting a new project.md:9]: If you're refering to the file name, then capitalize all words in the name, otherwise it looks like the project is called "Project"

File: Starting a new project.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+We've come up with a project setup template in cooperation with DevOps team. It's purpose is to provide the DevOps with all the details needed for deploying your application to staging and production servers.
```
Comment (cukoreva) [inline Starting a new project.md:11]:
```suggestion
We've come up with a project setup template in cooperation with the DevOps team. Its purpose is to provide the DevOps with all the details needed for deploying your application to staging and production servers.
```

File: Starting a new project.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+It contains general info about your project that your project manager can fill out as well as more technical information like frameworks and technologies used, build and deploy setup, third party services that need to be set up...
```
Comment (cukoreva) [inline Starting a new project.md:13]:
```suggestion
It contains general info about your project that your project manager can fill out, as well as more technical information like frameworks and technologies used, the build and deploy setup, third party services that need to be set up, etc.
```

File: Starting a new project.md (hunk 3 above)
Comment (cukoreva) [inline Starting a new project.md:17]:
```suggestion
**Fill in the blanks on backend-related stuff.**
```

File: Starting a new project.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+Ask your PM for a link to the sheet template.
+
+**Fill in the blanks on Backend related stuff.**
+
+If you need any other services that aren't mentioned, **provide as many information as possible**
```
Comment (cukoreva) [inline Starting a new project.md:19]:
```suggestion
If you need any other services that aren't mentioned, **provide as much information as possible**.
```

File: Starting a new project.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+**Project setup sheet needs to be up to date** with new changes to project setup and services used. If you're adding some new dependency or service or changing something, please update it in project setup sheet and notify DevOps to make these changes.
```
Comment (cukoreva) [inline Starting a new project.md:21]:
```suggestion
**The project setup sheet needs to be up to date** with new changes to the project setup and services used. If you're adding some new dependency or service or changing something, please update it in the project setup sheet and notify DevOps to make these changes.
```

File: Starting a new project.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+
+#### 3. Generate new application
+
+Use [**default rails template**](https://github.com/infinum/default_rails_template/) to start your new application.
```
Comment (cukoreva) [inline Starting a new project.md:26]:
```suggestion
Use the [**default rails template**](https://github.com/infinum/default_rails_template/) to start your new application.
```

File: Starting a new project.md [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+Most of our work is based on the Rails framework. We have some standard configurations and gems that are basically in every Rails application. To make our lives easier, we've made a default template to generate new Rails applications that meet those standards.
+
+
+#### 4. Check for new/better gems
+
+We usually use specific gems for some purposes, e.g. Devise for authentication, Pundit for authorization... That doesn't mean we necessarily need to use those gems.
```
Comment (cukoreva) [inline Starting a new project.md:33]:
```suggestion
We usually use specific gems for some purposes, e.g. Devise for authentication, Pundit for authorization, etc. That doesn't mean we need to use those gems necessarily.
```

File: Starting a new project.md [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+When checking out a gem, always look for these things:
+* is the gem maintained? when was the last commit?
+* how many stars does it have?
+* what is the ratio between open and closed issues?
+
+
+#### 5. Fill out readme
+
+After setting up a project, fill out a readme file with all the necessary details about the application:
```
Comment (cukoreva) [inline Starting a new project.md:45]:
```suggestion
After setting up a project, fill out the readme file with all the necessary details about the application:
```

File: Starting a new project.md [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+* setup instructions
+* development instructions
+* deployment instructions
+* test instructions
+* any other project specific information that someone who takes over a project needs to know
+
+
+#### 6. Create ER diagram
+
+Before you start coding, you always need to create an ER diagram for your database. Go through project specification and list all the entities needed together with their attributes and relationships between them.
```
Comment (cukoreva) [inline Starting a new project.md:56]:
```suggestion
Before you start coding, you always need to create an ER diagram for your database. Go through the project specification and list all the entities needed together with their attributes and relationships between them.
```

File: Starting a new project.md [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+For creating an ER diagram use:
+* [DB diagram](https://dbdiagram.io/)
+* [MySQL Workbench](https://dev.mysql.com/downloads/workbench/)
+
+After creating an ER diagram, you need to go through architecture review. You can read more about it in [Architecture review chapter](Architecture review)
```
Comment (cukoreva) [inline Starting a new project.md:62]:
```suggestion
After creating an ER diagram, you need to go through an architecture review. You can read more about it in the [Architecture review chapter](Architecture review)
```

File: Starting a new project.md [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+If you need to create an application or server architecture diagram use:
```
Comment (cukoreva) [inline Starting a new project.md:64]:
```suggestion
If you need to create an application or a server architecture diagram use:
```

File: Starting a new project.md (hunk 2 above)
Comment (d4be4st) [reply Starting a new project.md:51]: Does anybody use an ER diagram of an ongoing project?

It is either out of date (the one created before the coding started)
or it is created with railroady (or similar gem) and it can be a cluttered mess and unreadable.

I rather look at schema.db and follow the relalationships

Comment (lovro-bikic) [reply Starting a new project.md:51]: I think that keeping an ER diagram in sync with the schema creates overhead on ongoing projects because you have to remember to always update the diagram when you change the schema.

Also, [dbdiagram.io](https://dbdiagram.io) has an option to import schema.rb, so you can use that to quickly create a diagram if you'll ever need it:
image

Comment (cilim) [reply Starting a new project.md:51]: I like keeping the ERD in [repos](https://github.com/infinum/rails-messe-connect-networking#database-architecture) :)

It does sometimes get out of sync cause you forget or you just don't feel like updating it if you're adding a column or two...but every month or so it makes sense to me...

Comment (nikajukic) [reply Starting a new project.md:51]: I agree with @lovro-bikic. You'll always have to remember to keep it up to date and I think it's not efficient. If you're working on a project, you know it's architecture. And if you're new on the project you can check out schema and create a ER diagram with dbdiagram.io using schema.rb

Comment (vr4b4c) [reply Starting a new project.md:51]: Maintaining the ERD diagram is a subjective and opinionated topic
* which tools to use
* rate of updates
* allowed diff between DB schema and ERD diagram
* artifact location...

We should discuss this topic internally, come to an agreement if it's necessary or not. If we agree to maintain ERD diagrams in projects, we'll also define the mechanics of how to do it. At that point, we can circle back and update the handbook.

The bottom line, we should move this discussion to a more appropriate channel.

Comment (melcha) [reply Starting a new project.md:51]: Ok. @nikajukic bring up this topic on the next backend talks

Comment (nikajukic) [reply Starting a new project.md:51]: @melcha deal! :)

**Comment by melcha** [comment]
> The 2nd chapter (Fill out Project setup sheet) is really Infinum related, and this is a public chapter. Should we move it to the private repo or?

Let's keep it public, the content might be useful to all developers. I don't think it matter that much that the template is private.

@nikajukic you can merge this when you make the changes Eva's suggested.
